src/lib/client/abstract.ts

```diff
@@ -0,0 +1,132 @@
+import { ApiClientConfiguration, ApiVersion } from './client.model';
+import {
+  errorHandler,
+  fetchTransport,
+  Transport,
+  TransportResponse,
+} from '../transport';
+import { HEADERS } from '../constants';
+
+export abstract class ApiClient<Response, Params = undefined> {
+  protected abstract url: string;
+  protected abstract method: 'get' | 'post';
+  private static DEFAULT_CONFIG = {
+    host: 'https://api.what3words.com',
+    apiVersion: ApiVersion.Version3,
+  };
+  private _config: ApiClientConfiguration;
+  private transport: Transport;
+
+  constructor(
+    private _apiKey: string = '',
+    config: ApiClientConfiguration = {},
+    transport: Transport = fetchTransport()
+  ) {
+    this._config = Object.assign({}, ApiClient.DEFAULT_CONFIG, config);
+    this.transport = transport;
+  }
+
+  public apiKey(apiKey?: string): this | string {
+    if (apiKey !== undefined) {
+      this._apiKey = apiKey;
+      return this;
+    }
+    return this._apiKey;
```

Comment:
Line 32 returns the whole object whereas this line returns just the api key. Is this intentional, to have the two branches have different structures?

Reply:
Yes this is intentional. It allows you to chain set command such as

```
client.setApiKey('key')
 .setConfig({});
```

Whereas with get it returns the value so chaining wouldn't make sense.